src/ser/raw/value_serializer.rs

```diff
@@ -0,0 +1,542 @@
+use std::{convert::TryFrom, io::Write};
+
+use serde::{
+    ser::{Error as SerdeError, Impossible, SerializeMap, SerializeStruct},
+    Serialize,
+};
+
+#[cfg(feature = "decimal128")]
+use crate::Decimal128;
+use crate::{
+    oid::ObjectId,
+    ser::{write_binary, write_cstring, write_i32, write_i64, write_string, Error, Result},
+    spec::{BinarySubtype, ElementType},
+};
+
+use super::{document_serializer::DocumentSerializer, Serializer};
+
+/// A serializer used specifically for serializing the serde-data-model form of a BSON type (e.g.
+/// `Binary`) to raw bytes.
+pub(crate) struct ValueSerializer<'a> {
```

Comment:
This type is used to interpret any extended JSON formatted types in serde's data model and serialize them to actual BSON. It keeps track of where it's at via a state machine.